fv3core/decorators.py

```diff
@@ -149,7 +149,96 @@ def __iter__(self):
         return self.cache.__iter__()
 
 
-class FV3StencilObject:
+class StencilWrapper:
+    """Wrapped GT4Py stencil object."""
+
+    def __init__(
+        self,
+        func: Callable[..., None],
+        origin: Optional[Index3D] = None,
+        domain: Optional[Index3D] = None,
+        **kwargs,
+    ):
+        self.func = func
+        self.origin = origin
+        self.domain = domain
+
+        if "format_source" not in kwargs:
+            kwargs["format_source"] = global_config.get_format_source()
+
+        self.stencil_object = gtscript.stencil(
+            definition=self.func,
+            backend=global_config.get_backend(),
+            rebuild=global_config.get_rebuild(),
+            **kwargs,
+        )
+
+    def __call__(self, *args, **kwargs) -> None:
+        if self.origin:
+            assert "origin" not in kwargs, "Cannot override origin provided at init"
+            kwargs["origin"] = self.origin
+        kwargs["origin"] = self._get_field_origins(*args, **kwargs)
```

Comment:
I think we should store this information to only compute it once. 
```suggestion
        if not self.fields_with_origin:
           self.fields_with_origin = self._get_field_origins(*args, **kwargs)
[...]
elf.stencil_object.run(**kwargs, origin=self.fields_with_origin),
```

Reply:
Implemented, testing in progress...